src/sentry/runner/commands/run.py

```diff
@@ -702,36 +702,56 @@ def basic_consumer(consumer_name, consumer_args, topic, **options):
 
         sentry run consumer ingest-occurrences --consumer-group occurrence-consumer -- --help
     """
-    from sentry.consumers import KAFKA_CONSUMERS
+    from sentry.consumers import get_stream_processor
+    from sentry.metrics.middleware import add_global_tags
+    from sentry.utils.arroyo import initialize_arroyo_main
 
-    try:
-        consumer_definition = KAFKA_CONSUMERS[consumer_name]
-    except KeyError:
-        raise click.ClickException(
-            f"No consumer named {consumer_name} in sentry.consumers.KAFKA_CONSUMERS"
-        )
+    add_global_tags(kafka_topic=topic, consumer_group=options["group_id"])
+    initialize_arroyo_main()
 
-    try:
-        strategy_factory_cls = import_string(consumer_definition["strategy_factory"])
-        default_topic = consumer_definition["topic"]
-    except KeyError:
-        raise click.ClickException(
-            f"The consumer group {consumer_name} does not have a strategy factory"
-            f"registered. Most likely there is another subcommand in 'sentry run' "
-            f"responsible for this consumer"
+    processor = get_stream_processor(consumer_name, consumer_args, topic=topic, **options)
+    run_processor_with_signals(processor)
+
+
+@run.command("dev-consumer")
+@click.argument("consumer_names", nargs=-1)
+@log_options()
+@configuration
+def dev_consumer(consumer_names):
+    """
+    Launch multiple "new-style" consumers in the same thread.
+
+    This does the same thing as 'sentry run consumer', but is not configurable,
+    hardcodes consumer groups and is highly imperformant.
+    """
+
+    from sentry.consumers import get_stream_processor
+    from sentry.utils.arroyo import initialize_arroyo_main
+
+    initialize_arroyo_main()
+
+    processors = [
+        get_stream_processor(
+            consumer_name,
+            [],
+            topic=None,
+            group_id="sentry-consumer",
+            auto_offset_reset="latest",
+            strict_offset_reset=False,
         )
+        for consumer_name in consumer_names
+    ]
 
-    cmd = click.Command(
-        name=consumer_name, params=list(consumer_definition.get("click_options") or ())
-    )
-    cmd_context = cmd.make_context(consumer_name, list(consumer_args))
-    strategy_factory = cmd_context.invoke(
-        strategy_factory_cls, **cmd_context.params, **consumer_definition.get("static_args") or {}
-    )
+    def handler(signum, frame):
+        for processor in processors:
+            processor.signal_shutdown()
 
-    from sentry.utils.arroyo import run_basic_consumer
+    signal.signal(signal.SIGINT, handler)
+    signal.signal(signal.SIGTERM, handler)
 
-    run_basic_consumer(topic=topic or default_topic, **options, strategy_factory=strategy_factory)
+    while True:
+        for processor in processors:
+            processor._run_once()
```

Comment:
do I assume correctly that this indeed busy-loops when no messages are being received?
does `_run_once` return whether it handled a message? you could put in a `sleep` in case none of the processors received a message in one iteration?

Reply:
It will hang on `poll(timeout=1.0)` on the underlying rdkafka consumer here: https://github.com/getsentry/arroyo/blob/c8a79bce00fb079a0e7fc931ec5c878b7227fef7/arroyo/processing/processor.py#L355

so I see no chance of spinning.

what I am more concerned about is that the consumer spends 10 seconds polling for 10 consumers in order, and only then gets around to processing one message from the consumer we care about. So the throughput and latency might be too low even for development purposes. but I don't see another way right now, because an arroyo consumer can't be launched in a separate thread

Reply:
having to wait ~10 seconds for one round of messages is indeed very limiting :-(

Does it at least process more than one message on a single `_run_once()`? For things like minidumps, it would otherwise take ~20s to process a single event… ?

Reply:
you're right and I see no way of fixing that right now. I would suggest that if you care about this, you just don't use `--dev-consumer` (which is opt-in) or you spawn a separate ingest-attachments consumer to speed up the process